[4.1] mod_breadcrumbs - convert microdata to JSON-LD

hans2103 opened the pull request:
Pull Request for Issue # .

### Summary of Changes

With https://github.com/joomla/joomla-cms/pull/25117 in mind and the possible adaptation of Spatie\SchemaOrg in mind I would like to prepare current breadcrumbs by converting the Microdata into JSON-LD structure. 
But not only with that possible adaptation in mind... The JSON-LD format is the recommended format. 
The HTML code will be cleaner when the data is moved to JSON-LD format
Another benefit of this change is that the current page will always be added to the data. This is not the case in the current situation with _Hide Last Item_ is active. 


### Testing Instructions

- Install Joomla with Sample Data Blog. 
- Note that there will be a  mod_breadcrumbs on every page. example: http://**your-joomla-4-site**/index.php/blog/article
- Inspect element and see the markup of mod_breadcrumbs
- Apply this patch. No `npm run something:something` is required
- Refresh the page and note the changed markup.


### Actual result BEFORE applying this Pull Request

```
<nav role="navigation" aria-label="Breadcrumbs">
    <ol itemscope="" itemtype="https://schema.org/BreadcrumbList" class="mod-breadcrumbs breadcrumb">
        <li class="mod-breadcrumbs__here float-left">You are here: &nbsp;</li>
        <li itemprop="itemListElement" itemscope="" itemtype="https://schema.org/ListItem" class="mod-breadcrumbs__item breadcrumb-item">
            <a itemprop="item" href="/index.php" class="pathway"><span itemprop="name">Home</span></a>
            <meta itemprop="position" content="1">
        </li>
        <li itemprop="itemListElement" itemscope="" itemtype="https://schema.org/ListItem" class="mod-breadcrumbs__item breadcrumb-item">
            <a itemprop="item" href="/index.php/blog" class="pathway"><span itemprop="name">Blog</span></a>
            <meta itemprop="position" content="2">
        </li>
        <li aria-current="page" itemprop="itemListElement" itemscope="" itemtype="https://schema.org/ListItem" class="mod-breadcrumbs__item breadcrumb-item active">
            <span itemprop="name">Article</span>
            <meta itemprop="position" content="3">
        </li>
    </ol>
</nav>

```


### Expected result AFTER applying this Pull Request

```
<nav role="navigation" aria-label="Breadcrumbs">
    <ol class="mod-breadcrumbs breadcrumb">
        <li class="mod-breadcrumbs__here float-left">You are here: &nbsp;</li>
        <li class="mod-breadcrumbs__item breadcrumb-item"><a href="/index.php" class="pathway"><span>Home</span></a></li>
        <li class="mod-breadcrumbs__item breadcrumb-item"><a href="/index.php/blog" class="pathway"><span>Blog</span></a></li>
        <li class="mod-breadcrumbs__item breadcrumb-item active"><span>Article</span></li>
    </ol>
</nav>
```

While inspecting the element and scroll to top and inspect element `<head>`. It should contain the following `<script>` tag. 
```
    <script type="application/ld+json">
        {
            "@context": "https:\/\/schema.org",
            "@type": "BreadcrumbList",
            "itemListElement": [
                {
                    "@type": "ListItem",
                    "position": 1,
                    "item": {
                        "@id": "http:\/\/joomla4.test\/index.php",
                        "name": "Home"
                    }
                },
                {
                    "@type": "ListItem",
                    "position": 2,
                    "item": {
                        "@id": "http:\/\/joomla4.test\/index.php\/blog",
                        "name": "Blog"
                    }
                },
                {
                    "@type": "ListItem",
                    "position": 3,
                    "item": {
                        "@id": "http:\/\/joomla4.test\/index.php\/blog\/article",
                        "name": "Article"
                    }
                }
            ]
        }    </script>
```


### Documentation Changes Required

hans2103 committed:
:recycle: convert microdata to JSON-LD

HLeithner commented:
We are in feature freeze for j4 please rebase to j4.1

hans2103 commented:
@HLeithner can you help me how I rebase to 4.1?

HLeithner commented:
You can try to just target the 4.1-dev branch in github (you can edit this PR and select another target branch).

Depending on how long it takes to merge and if 4.1-dev is uptodate it could work.

Else: https://git-scm.com/book/en/v2/Git-Branching-Rebasing

But it's maybe easier to create a new branch of the 4.1 branch and copy paste.

hans2103 commented:
changed base.... already checked that mod_breadcrumbs/tmpl/default.php J4 = J4.1

richard67 commented:
Bad news: Because the 4.1-dev branch is not up to date with latest changes in the 4.0-dev branch, this PR now shows all changes in 4.0-dev since the 4.1-dev was updated to 4.0-dev last time, and not only the changes from this PR.

Maybe that will be fixed by just waiting for the next update of the 4.1-dev branch from 4.0-dev, but I'm not sure about that.

HLeithner commented:
@zero-24 or @wilsonge  can you please update the branch, I'm unable at the moment to do this.

HLeithner commented:
@hans2103 keep it as it is for the moment.

ceford commented:
Using Patchtester I get this:

The file marked for modification does not exist: administrator/components/com_media/resources/scripts/components/toolbar/toolbar.vue

Is this to do with the base change?
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30727">issues.joomla.org/tracker/joomla-cms/30727</a>.</sub>

richard67 commented:
@ceford Yes, you should wait until the 4.1-dev branch is up to date, and then you can test this PR only on a 4.1 and not on a 4.0 installation.

wilsonge commented:
Should be good now - pushed it up into 4.1-dev (needed to swap the branches around to make github realise)

SharkyKZ commented on modules/mod_breadcrumbs/tmpl/default.php:
```diff
@@ -36,32 +38,56 @@
+		];
+	}
+	?>
+	<script type="application/ld+json">
```
Use Document API to add scripts.

hans2103 replied:
@SharkyKZ How to use Document API to add `application/ld+json` ?

SharkyKZ replied:
$app->getDocument()->addScriptDeclaration($data, 'application/ld+json');

Or using web asset:

    $app->getDocument()->getWebAssetManager()->addInline('script', $data, [], ['type' => 'application/ld+json']);

hans2103 replied:
done with commit https://github.com/joomla/joomla-cms/pull/30727/commits/2fbe114a8d3860bef84f8ec58552691c3a721fe0

SharkyKZ commented on modules/mod_breadcrumbs/tmpl/default.php:
```diff
@@ -36,32 +38,56 @@
+				'@type'    => 'ListItem',
+				'position' => $key + 1,
+				'item'     => [
+						'@id'  => $item->link ? Uri::base() . ltrim($item->link, '/') : Uri::current(),
```
This results in broken external links. Use `Joomla\CMS\Router\Route::_()` instead..

SharkyKZ replied:
It seems it's already used in the helper. Should be moved to the layout. Will make a PR.

hans2103 replied:
thank you. please refer your PR here

hans2103 replied:
merged recent updates on 4.1 into my repo and found out that you already solved it. Applied changes

richard67 commented:
Drone seems to have a problem. I think when adding a change (e.g. those requested by review above) it will be fixed.

joomlacorner commented:
I have tested this item :white_check_mark: successfully on 4710faf347984ce27573c7a61cfc14112978b6d3<br /><br />Worked as Expected result.


'''
<nav role="navigation" aria-label="Breadcrumbs">
	<ol class="mod-breadcrumbs breadcrumb">
					<li class="mod-breadcrumbs__here float-left">
				You are here: &#160;
			</li>
		
		<li class="mod-breadcrumbs__item breadcrumb-item"><a href="/index.php" class="pathway"><span>Home</span></a></li><li class="mod-breadcrumbs__item breadcrumb-item"><a href="/index.php/blog" class="pathway"><span>Blog</span></a></li><li class="mod-breadcrumbs__item breadcrumb-item active"><span>Welcome to your blog</span></li>	</ol>
		<script type="application/ld+json">
	{"@context":"https:\/\/schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"item":{"@id":"https:\/\/j4.akarawuth.com\/index.php","name":"Home"}},{"@type":"ListItem","position":2,"item":{"@id":"https:\/\/j4.akarawuth.com\/index.php\/blog","name":"Blog"}},{"@type":"ListItem","position":3,"item":{"@id":"https:\/\/j4.akarawuth.com\/index.php\/blog\/3-welcome-to-your-blog","name":"Welcome to your blog"}}]}	</script>
</nav>
'''
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30727">issues.joomla.org/tracker/joomla-cms/30727</a>.</sub>

hans2103 committed:
:recycle: move script to WebAssetManager

hans2103 committed:
Merge branch '4.1-dev' into feature/mod_breadcrumbs

# Conflicts:
#	modules/mod_breadcrumbs/tmpl/default.php

SharkyKZ commented on modules/mod_breadcrumbs/tmpl/default.php:
```diff
@@ -9,10 +9,13 @@
 use Joomla\CMS\Uri\Uri;
+use Joomla\CMS\WebAsset\WebAssetManager;
 
+$app       = Factory::getApplication();
```
```suggestion
```

SharkyKZ commented on modules/mod_breadcrumbs/tmpl/default.php:
```diff
@@ -9,10 +9,13 @@
 
 defined('_JEXEC') or die;
 
+use Joomla\CMS\Factory;
```
```suggestion
```

hans2103 committed:
Update modules/mod_breadcrumbs/tmpl/default.php

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

hans2103 committed:
Update modules/mod_breadcrumbs/tmpl/default.php

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

SharkyKZ commented on modules/mod_breadcrumbs/tmpl/default.php:
```diff
@@ -37,32 +40,57 @@
 		$show_last = $params->get('showLast', 1);
 
+		$class   = null;
+		$divider = null;
```
`$divider` is not used.

hans2103 replied:
removed in commit https://github.com/joomla/joomla-cms/pull/30727/commits/4a863cf050ce26108202e811eceef7d548cff1e9

hans2103 committed:
removal non used variable

SharkyKZ commented on modules/mod_breadcrumbs/tmpl/default.php:
```diff
@@ -37,32 +40,56 @@
+				'@type'    => 'ListItem',
+				'position' => $key + 1,
+				'item'     => [
+						'@id'  => $item->link ? Route::_(Uri::base() . ltrim($item->link, '/')) : Route::_(Uri::current()),
```
```suggestion
						'@id'  => $item->link ? Route::_($item->link, true, Route::TLS_IGNORE, true) : Route::_(Uri::current()),
```

SharkyKZ replied:
What is the purpose of using `Uri::current()` here?

hans2103 replied:
If, for some reason the item->link cannot be created, Joomla will provide the url of the current page. That was the idea behind this.

SharkyKZ replied:
Well, if you need the exact page, use `Uri::getInstance()` because `Uri::current()` strips out query strings. And don't need  to pass this to `Route::_()`

hans2103 replied:
thank you... done with commit https://github.com/joomla/joomla-cms/pull/30727/commits/f38b735ea76b0a55f2363d4856a8f221805fecb2

hans2103 committed:
Update modules/mod_breadcrumbs/tmpl/default.php

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

hans2103 committed:
replace Uri::current() with Uri::getInstance()

hans2103 committed:
Merge branch '4.1-dev' into feature/mod_breadcrumbs

adj9 commented:
I have tested this item :white_check_mark: successfully on fdf12e3060b48e31462da7115c372c86351f7987<br /><br />done :)!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30727">issues.joomla.org/tracker/joomla-cms/30727</a>.</sub>

gostn commented:
HTML-Markup successfully, didnt found the script-part:

![image](https://user-images.githubusercontent.com/71960554/97100556-d2a60a80-1694-11eb-8814-47591dad7f13.png)

hans2103 commented:
@gostn hve you checked the '<head>' element?

gostn commented:
no idea what you mean, i followed your test instructions.

hans2103 commented:
@gostn I see... 
with commit https://github.com/joomla/joomla-cms/commit/2fbe114a8d3860bef84f8ec58552691c3a721fe0 I have moved the json block to the `<head>` of the document. Suggested by @SharkyKZ 

Just adjusted the test instructions. Can you test again?

gostn commented:
I have tested this item :white_check_mark: successfully on fdf12e3060b48e31462da7115c372c86351f7987<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30727">issues.joomla.org/tracker/joomla-cms/30727</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30727">issues.joomla.org/tracker/joomla-cms/30727</a>.</sub>
